Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1739 @@
+homomorphism."]
+def invMonoidHom : α →* α where
+  toFun := Inv.inv
+  map_one' := inv_one
+  map_mul' := mul_inv
+#align inv_monoid_hom invMonoidHom
+
+@[simp]
+theorem coe_invMonoidHom : (invMonoidHom : α → α) = Inv.inv := rfl
+#align coe_inv_monoid_hom coe_invMonoidHom
+
+@[simp]
+theorem invMonoidHom_apply (a : α) : invMonoidHom a = a⁻¹ := rfl
+#align inv_monoid_hom_apply invMonoidHom_apply
+
+end DivisionCommMonoid
+
+/-- The identity map from a type with 1 to itself. -/
+@[to_additive, simps]
+def OneHom.id (M : Type _) [One M] : OneHom M M where
+  toFun x := x
+  map_one' := rfl
+#align one_hom.id OneHom.id
+
+/-- The identity map from a type with multiplication to itself. -/
+@[to_additive, simps]
+def MulHom.id (M : Type _) [Mul M] : M →ₙ* M where
+  toFun x := x
+  map_mul' _ _ := rfl
+#align mul_hom.id MulHom.id
+
+/-- The identity map from a monoid to itself. -/
+@[to_additive, simps]
+def MonoidHom.id (M : Type _) [MulOneClass M] : M →* M where
+  toFun x := x
+  map_one' := rfl
+  map_mul' _ _ := rfl
+#align monoid_hom.id MonoidHom.id
+
+/-- The identity map from a monoid_with_zero to itself. -/
+@[simps]
+def MonoidWithZeroHom.id (M : Type _) [MulZeroOneClass M] : M →*₀ M where
+  toFun x := x
+  map_zero' := rfl
+  map_one' := rfl
+  map_mul' _ _ := rfl
+#align monoid_with_zero_hom.id MonoidWithZeroHom.id
+
+/-- The identity map from an type with zero to itself. -/
+add_decl_doc ZeroHom.id
+
+/-- The identity map from an type with addition to itself. -/
+add_decl_doc AddHom.id
+
+/-- The identity map from an additive monoid to itself. -/
+add_decl_doc AddMonoidHom.id
+
+/-- Composition of `OneHom`s as a `OneHom`. -/
+@[to_additive]
+def OneHom.comp [One M] [One N] [One P] (hnp : OneHom N P) (hmn : OneHom M N) : OneHom M P where
+  toFun := hnp ∘ hmn
+  map_one' := by simp
+#align one_hom.comp OneHom.comp
+
+/-- Composition of `MulHom`s as a `MulHom`. -/
+@[to_additive]
+def MulHom.comp [Mul M] [Mul N] [Mul P] (hnp : N →ₙ* P) (hmn : M →ₙ* N) : M →ₙ* P where
+  toFun := hnp ∘ hmn
+  map_mul' x y := by simp
+#align mul_hom.comp MulHom.comp
+
+-- Porting note: restore `by simp`
+/-- Composition of monoid morphisms as a monoid morphism. -/
+@[to_additive]
+def MonoidHom.comp [MulOneClass M] [MulOneClass N] [MulOneClass P] (hnp : N →* P) (hmn : M →* N) :
+  M →* P where
+  toFun := hnp ∘ hmn
+  map_one' := by rw [Function.comp_apply, map_one, map_one]
+  map_mul' := by simp
+#align monoid_hom.comp MonoidHom.comp
+
+-- Porting note: something strange in `map_one'`, where the goal is `↑1 = 1` via
+-- `MulZeroOneClass.toZero`
+/-- Composition of `MonoidWithZeroHom`s as a `MonoidWithZeroHom`. -/
+def MonoidWithZeroHom.comp [MulZeroOneClass M] [MulZeroOneClass N] [MulZeroOneClass P]
+  (hnp : N →*₀ P) (hmn : M →*₀ N) : M →*₀ P where
+  toFun := hnp ∘ hmn
+  map_zero' := by rw [Function.comp_apply, map_zero, map_zero]
+  map_one' := by dsimp only []; rw [Function.comp_apply, map_one, map_one]
```

Comment:
The goal presented in `map_one'` is the cryptic `↑1 = 1`. This problem is in addition to the `map_one`/`simp` problem above.